flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionDeployer.java

```diff
@@ -0,0 +1,386 @@
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Default implementation of {@link ExecutionDeployer}. */
+public class DefaultExecutionDeployer implements ExecutionDeployer {
+
+    private final Logger log;
+
+    private final ExecutionSlotAllocator executionSlotAllocator;
+
+    private final ExecutionOperations executionOperations;
+
+    private final ExecutionVertexVersioner executionVertexVersioner;
+
+    private final Time partitionRegistrationTimeout;
+
+    private final Function<ExecutionAttemptID, Execution> executionRetriever;
+
+    private final BiConsumer<ExecutionVertexID, AllocationID> allocationReservationFunc;
+
+    private final ComponentMainThreadExecutor mainThreadExecutor;
+
+    private DefaultExecutionDeployer(
+            final Logger log,
+            final ExecutionSlotAllocator executionSlotAllocator,
+            final ExecutionOperations executionOperations,
+            final ExecutionVertexVersioner executionVertexVersioner,
+            final Time partitionRegistrationTimeout,
+            final Function<ExecutionAttemptID, Execution> executionRetriever,
+            final BiConsumer<ExecutionVertexID, AllocationID> allocationReservationFunc,
+            final ComponentMainThreadExecutor mainThreadExecutor) {
+
+        this.log = checkNotNull(log);
+        this.executionSlotAllocator = checkNotNull(executionSlotAllocator);
+        this.executionOperations = checkNotNull(executionOperations);
+        this.executionVertexVersioner = checkNotNull(executionVertexVersioner);
+        this.partitionRegistrationTimeout = checkNotNull(partitionRegistrationTimeout);
+        this.executionRetriever = checkNotNull(executionRetriever);
+        this.allocationReservationFunc = checkNotNull(allocationReservationFunc);
+        this.mainThreadExecutor = checkNotNull(mainThreadExecutor);
+    }
+
+    @Override
+    public void allocateSlotsAndDeploy(
+            final List<Execution> executionsToDeploy,
+            final Map<ExecutionVertexID, ExecutionVertexVersion> requiredVersionByVertex) {
+        validateExecutionStates(executionsToDeploy);
+
+        transitionToScheduled(executionsToDeploy);
+
+        final List<ExecutionSlotAssignment> slotExecutionVertexAssignments =
+                allocateSlotsFor(executionsToDeploy);
+
+        final List<ExecutionDeploymentHandle> deploymentHandles =
+                createDeploymentHandles(requiredVersionByVertex, slotExecutionVertexAssignments);
+
+        waitForAllSlotsAndDeploy(deploymentHandles);
+    }
+
+    private void validateExecutionStates(final Collection<Execution> executionsToDeploy) {
+        executionsToDeploy.stream()
+                .forEach(
```

Comment:
```suggestion
        executionsToDeploy.forEach(
```

Reply:
Maybe _Iterable#forEach_ is more concise.

Reply:
You are right.